app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/hplus/HPlusHandlerThread.java

```diff
@@ -0,0 +1,446 @@
+            HPlusHealthSampleProvider provider = new HPlusHealthSampleProvider(getDevice(), dbHandler.getDaoSession());
+
+            //Insert the Overlays
+            List<HPlusHealthActivityOverlay> overlayList = new ArrayList<>();
+            List<HPlusDataRecord.RecordInterval> intervals = record.getIntervals();
+            for(HPlusDataRecord.RecordInterval interval : intervals){
+                overlayList.add(new HPlusHealthActivityOverlay(interval.timestampFrom, interval.timestampTo, interval.activityKind, deviceId, userId, null));
+            }
+
+            overlayDao.insertOrReplaceInTx(overlayList);
+
+            //Store the data
+            HPlusHealthActivitySample sample = new HPlusHealthActivitySample(
+                    record.timestamp,            // ts
+                    deviceId, userId,            // User id
+                    record.getRawData(),         // Raw Data
+                    record.activityKind,
+                    0,                           // Intensity
+                    ActivitySample.NOT_MEASURED, // Steps
+                    ActivitySample.NOT_MEASURED, // HR
+                    ActivitySample.NOT_MEASURED, // Distance
+                    ActivitySample.NOT_MEASURED  // Calories
+            );
+
+            sample.setProvider(provider);
+
+            provider.addGBActivitySample(sample);
+
+
+        } catch (Exception ex) {
+            LOG.debug(ex.getMessage());
+        }
+
+        mGetSleepTime = Calendar.getInstance();
+        mGetSleepTime.add(Calendar.SECOND, SLEEP_SYNC_PERIOD);
+
+        return true;
+    }
+
+    private void requestNextSleepData() {
+        mGetSleepTime = Calendar.getInstance();
+        mGetSleepTime.add(Calendar.SECOND, SLEEP_RETRY_PERIOD);
+
+        TransactionBuilder builder = new TransactionBuilder("requestSleepStats");
+        builder.write(mHPlusSupport.ctrlCharacteristic, new byte[]{HPlusConstants.CMD_GET_SLEEP});
+        builder.queue(mHPlusSupport.getQueue());
+    }
+
+
+    public boolean processRealtimeStats(byte[] data) {
+        HPlusDataRecordRealtime record;
+
+        try{
+            record = new HPlusDataRecordRealtime(data);
+        } catch(IllegalArgumentException e){
+            LOG.debug((e.getMessage()));
+            return true;
+        }
+
+        if(record.same(prevRealTimeRecord))
+            return true;
+
+        prevRealTimeRecord = record;
+
+        getDevice().setBatteryLevel(record.battery);
+        getDevice().sendDeviceUpdateIntent(getContext());
+
+        //Skip when measuring
+        if(record.heartRate == 255) {
+            getDevice().setFirmwareVersion2("---");
+            getDevice().sendDeviceUpdateIntent(getContext());
+            return true;
+        }
+
+        getDevice().setFirmwareVersion2("" + record.heartRate);
+        getDevice().sendDeviceUpdateIntent(getContext());
+
+        try (DBHandler dbHandler = GBApplication.acquireDB()) {
```

Comment:
Please correct me :)

The device will report Heart Rate periodically. Each message is independent of the previous messages, so I do not see that we reach an end. Also, this data is not stored in the device. Either we store it or it can be lost. I already filter data so that we only store unique packets.

Day and Sleep summaries are stored in the device. However, they are only updated once per day (each).

Day details are sent in slots, with 144 slots per day. This can be buffered and then stored.